### ISSUE TITLE
Ensembles of minimizers

### ISSUE THREAD

**Issue by betatim** [scikit-optimize/scikit-optimize#195]
Ensembling/stacking of classifiers seems to be the winning strategy for most kaggle competitions.

It would be interesting to investigate how to construct an "ensemble" of minimizers and then evaluate if the ensemble performs better than an individual minimizer.

The ensemble could be formed by combining predictions with a model (a la [VotingClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.ensemble.VotingClassifier.html)) or maybe use a library: http://www.niculescu-mizil.org/papers/shotgun.icml04.revised.rev2.pdf

You could try different kernels and different methods/hyper-parameters for each method as the constituents of the ensemble.

Not clear to me that this would improve things, but sounds interesting if someone wants to do some research.

Would need good benchmarks: #180

**Comment by betatim** [comment]
At each step of the optimisation problem have a set of surrogate models and treat choosing between them as a multi-armed bandit problem (similar to GP hedge).

Use a GP based surrogate at the start and then switch to a tree based on after you have collected "many" points. Idea is to use a GP when you have little data and trees when runtime gets slow when you have many points.

**Comment by iaroslav-ai** [comment]
hmm interesting. Right now this is actually something that is already happening - for some n_random_starts iterations, we are using dummy_minimize, and then gp. Could be extended somewhere in that direction. 

My experience so far with ensembling is that stacking works best (train a single model on outputs of multiple models to combine them together) but mostly on larger sized datasets. So maybe use something like this, but with n_calls being "large". Of course I might be biased on problems of my own.

**Comment by betatim** [comment]
Clarifying [previous post](https://github.com/scikit-optimize/scikit-optimize/issues/195#issuecomment-323506873) (because I struggled to remember what I meant): these are two distinct ideas for "ensembles".

@iaroslav-ai so you think something worth trying would be to train a GP, a GBT, RF, etc based surrogate model and then stack those? We could also use one kind of model with many different hyper-parameters? This would be a poor man's "integrate out surrogate model's hyper-parameters" I guess...

**Comment by iaroslav-ai** [comment]
Yes, right, I somehow drifted from one topic to another without any clear transition :)

Yes I would say so. I like both ideas that you propose. I would be quite interested how such things would behave on benchmarks. 

There are some interesting works on ensembling the surrogate models:

[https://scirate.com/arxiv/1706.01445](https://scirate.com/arxiv/1706.01445) : here ensembling somehow is used to speed up BO for problems with up to 10000 evaluations by means of paralellization.

[https://link.springer.com/article/10.1007/s10898-014-0184-0](https://link.springer.com/article/10.1007/s10898-014-0184-0) Seems to be close in spirit to stacking. They say that ensembles of surrogates worked best for their problems. I cannot get open access to the article though :/
 
Maybe some ensembling ideas can be borrowed from auto sklearn [https://papers.nips.cc/paper/5872-efficient-and-robust-automated-machine-learning](https://papers.nips.cc/paper/5872-efficient-and-robust-automated-machine-learning). I remember there they discussed ensembling in detail, and found some algorithm to work best other then stacking - there was some overfitting going on with latter.  

I think this could end up with some interesting novel things.
